src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensCache.cs

```diff
@@ -0,0 +1,136 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Roslyn.Utilities;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.SemanticTokens
+{
+    /// <summary>
+    /// Caches the semantic token information that needs to be preserved between multiple
+    /// semantic token requests.
+    /// Multiple token sets can be cached per document. The number of token sets cached
+    /// per document is determined by the _maxCachesPerDoc field.
+    /// </summary>
+    [Export(typeof(SemanticTokensCache)), Shared]
+    internal class SemanticTokensCache
+    {
+        /// <summary>
+        /// Maps a LSP token type to its respective index recognized by LSP.
+        /// </summary>
+        public static readonly Dictionary<string, int> TokenTypeToIndex;
+
+        /// <summary>
+        /// Number of cached token sets we store per document. Must be >= 1.
+        /// </summary>
+        private readonly int _maxCachesPerDoc = 5;
+
+        /// <summary>
+        /// The next resultId available to use. Atomically incremented with Interlocked,
+        /// so this doesn't need to be protected by _semaphore.
+        /// </summary>
+        private long _nextResultId;
+
+        /// <summary>
+        /// Multiple cache requests or updates may be received concurrently.
+        /// We need this sempahore to ensure that we aren't making concurrent
+        /// modifications to the _tokens dictionary.
+        /// </summary>
+        private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1);
+
+        #region protected by _semaphore
+        /// <summary>
+        /// Maps a document URI to its n most recently cached token sets.
+        /// </summary>
+        private readonly Dictionary<Uri, List<LSP.SemanticTokens>> _tokens =
+            new Dictionary<Uri, List<LSP.SemanticTokens>>();
+        #endregion
+
+        static SemanticTokensCache()
+        {
+            // Computes the mapping between a LSP token type and its respective index recognized by LSP.
```

Comment:
can you clarify "respective index recognized by lsp".  i'm not quite sure what that means.

Reply:
I modified the comment, I agree it was confusing before. Hopefully this is clearer:
```
        /// Maps a LSP token type to the index LSP associates with the token.
       /// This is needed since we need to report tokens back to LSP as ints,
       /// and LSP needs a way to decipher them.
```